### PR TITLE
[Project Euler] Added type hints and refactored the code a bit for Problem 14

### PR DESCRIPTION
**Pull request by sgalpha01**
Continuing #2786

- Added type hints
- Added default argument values for validate_solutions script
- Refactored the code

### **Describe your change:**

* [x] Fix a bug or typo in an existing algorithm?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).

### CODE REVIEW

File: project_euler/problem_14/sol1.py [hunk 1]
```diff
@@ -55,7 +55,10 @@ def solution(n):
 
 
 if __name__ == "__main__":
-    result = solution(int(input().strip()))
+    n = int(input().strip())
+    result = solution(n)
     print(
-        ("Largest Number:", result["largest_number"], "->", result["counter"], "digits")
+        "Longest Collatz sequence under {} is produced by {} with length {}".format(
```
Comment (dhruvmanila) [inline project_euler/problem_14/sol1.py:61]: Nice idea.
Please use `f-strings` as mentioned in CONTRIBUTING.md

File: project_euler/problem_14/sol2.py [hunk 2]
```diff
@@ -56,8 +56,10 @@ def solution(n):
 
 
 if __name__ == "__main__":
-    result = solution(int(input().strip()))
+    n = int(input().strip())
+    result = solution(n)
     print(
-        "Longest Collatz sequence under one million is %d with length %d"
-        % (result["largest_number"], result["counter"])
+        "Longest Collatz sequence under {} is produced by {} with length {}".format(
```
Comment (dhruvmanila) [inline project_euler/problem_14/sol2.py:62]: `f-strings`

File: project_euler/problem_14/sol1.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 """
 
 
-def solution(n):
+def solution(n: int = 1000000) -> dict:
```
Comment (dhruvmanila) [inline project_euler/problem_14/sol1.py:20]: Use `from typing import Dict` and then `Dict[<type>, <type>,...]`

File: project_euler/problem_14/sol2.py [hunk 4]
```diff
@@ -25,7 +25,7 @@
 """
 
 
-def collatz_sequence(n):
+def collatz_sequence(n: int) -> list:
```
Comment (dhruvmanila) [inline project_euler/problem_14/sol2.py:28]: Use `from typing import List` and then `List[<type>,...]`

File: project_euler/problem_14/sol2.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ def collatz_sequence(n):
     return sequence
 
 
-def solution(n):
+def solution(n: int = 1000000) -> dict:
```
Comment (dhruvmanila) [inline project_euler/problem_14/sol2.py:40]: Same as above.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

**Comment by sgalpha01** [comment]
Hey @dhruvmanila, thanks for reviewing. I have pushed some changes. Can you please check if it still need some styling modification?

**Comment by dhruvmanila** [comment]
This gives us answer in an incorrect type. 
https://travis-ci.com/github/TheAlgorithms/Python/jobs/396914178#L250-L256

**Comment by sgalpha01** [comment]
Sure!

**Review by dhruvmanila — APPROVED** [review]

**Pull request merged**